Use `aerich` for database migrations

bigherc18 opened the pull request:
This PR is about using `aerich` to add support for database migrations

Here's a demo: https://asciinema.org/a/Sw7OwQZKmWyznV3xgohjd59OT

I added a new config called `migrations_dir` to specify where migration files should be stored

```yaml [dipdup.yaml]
database:
  migrations_dir: ./migrations
```

A a set of commands have been added to manage database migrations:

- `dipdup schema history` to show the migration history
- `dipdup schema heads` to show the show the current available heads
- `dipdup schema migrate` to create a new migration changes file
- `dipdup schema upgrade` to upgrade the database to the latest or a specified version
- `dipdup schema downgrade` to downgrade the database to a previous version

Theses commands invoke `aerich`'s commands under the hood

The migrations directory should be initialized before using any of these commands, either with `dipdup schema init` or by running the indexer with `dipdup run`.

Note that `aerich` commands can't be used directly, as it expects the config to be stored in `pyproject.toml`

bigherc18 committed:
Draft: use aerich for database migrations

bigherc18 committed:
Fix database migration with aerich

- Write Changelog and a first draft for the docs
- Abort migration commands if it's not initialized
- Refactor the initialization to `DipDup._initialize_migrations`
- Schema is approved automatically after upgrade and downgrade
- `wipe_schema` also removes the `migration_dir`

droserasprout commented on src/dipdup/cli.py:
```diff
@@ -17,6 +17,7 @@
 
 import click
 import uvloop
+from aerich.cli import cli as aerich_cli
```
This import should be lazy, since it takes 0.6s and  `dipdup.cli` module is imported for every command.

droserasprout replied:
This import should be lazy, since it takes 0.6s and dipdup.cli module is imported for every command.

droserasprout commented on src/dipdup/cli.py:
```diff
@@ -562,7 +580,52 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
-    pass
+    from pathlib import Path
+
+    from dipdup.aerich import create_aerich_command
```
This import is not lazy enough. You could do it right before the call.

droserasprout commented on src/dipdup/config/__init__.py:
```diff
@@ -98,11 +99,13 @@ class SqliteDatabaseConfig:
     :param kind: always 'sqlite'
     :param path: Path to .sqlite file, leave default for in-memory database (`:memory:`)
     :param immune_tables: List of tables to preserve during reindexing
+    :param migrations_dir: Where to store migrations files (relative to project root)
```
I suggest removing an ability to configure this path and hardcode it in `DipDupPackage` class instead (but don't create by default). We already use fixed directory paths for every other section; it makes things much simpler to maintain.

droserasprout commented on pyproject.toml:
```diff
@@ -80,6 +80,7 @@ dependencies = [
     "tortoise-orm==0.21.6",
     "uvloop~=0.20",
     "web3~=7.2",
+    "aerich~=0.7.2",
```
Let's make this dependency optional. A majority of users don't need migrations, and dependency tree is already very bloated. If it's not installed -- just ignore the whole migrations part.

droserasprout commented on src/dipdup/cli.py:
```diff
@@ -562,7 +580,52 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
+    )
+
+
+schema.add_command(aerich_cli.commands['history'])
```
Please wrap this with try except to silently fail when aerich is not installed.

droserasprout reviewed (changes requested):
Hi! Thanks for your contribution! It's a great code with many smart solutions; I like every piece of it! My only concern is making this integration optional in every way: dependencies, imports, files in the project package.

bigherc18 committed:
Make database migrations optional in every way possible

- aerich is now an optional dependency under the optional migrations group
- migrations folder is hardcoded in the DipDupPackage class
- migrations commands and folder will only be added if aerich is installed

bigherc18 committed:
Make sure we catch only the exceptions we're expecting in _initialize_migrations

bigherc18 commented:
Do you guys prefer `rebase` or `merge` with `next` to fix the conflicts ?

droserasprout commented:
@bigherc18 We squash every merge request, so it's up to you what's inside.

bigherc18 commented:
Some concerns and limitations I have:

- `aerich` will be confused if the user changes the database. If we want to support changing the database, we have to make aerich aware of which database is being used.
- If you answer "No" to the question "Downgrade is dangerous, which may cause data loss. Are you sure? [y/N]" after running `dipdup schema downgrade`, the CLI hangs there forever.
- `aerich` has confusing messages after running `heads` and `history`. It makes the user feel like they have to take action when everything is actually fine: "No available heads, try migrate first."

droserasprout commented on .vscode/settings.json:
```diff
@@ -7,5 +7,8 @@
-    }
-}
\ No newline at end of file
+    },
```
Please, remove this part.

droserasprout commented on src/dipdup/database.py:
```diff
@@ -69,11 +69,14 @@ async def tortoise_wrapper(
 
     model_modules: dict[str, Iterable[str | ModuleType]] = {
         'int_models': ['dipdup.models'],
+        'models': ['dipdup.models'],
```
This line is a duplicate. Without it mypy seems to be okay with `model_modules['models'] = [models]`

droserasprout commented on src/dipdup/cli.py:
```diff
@@ -562,7 +579,59 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
+
+        if not migrations_dir.exists():
+            echo(
+                f"""Database migrations is not initialize at {migrations_dir}.
```
*are not initialized

droserasprout commented on src/dipdup/cli.py:
```diff
@@ -562,7 +579,59 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
+                f"""Database migrations is not initialize at {migrations_dir}.
+                 Run `dipdup schema init` or just run `dipdup run` and it'll be initialized automatically."""
+            )
+            raise click.Abort()
```
Please, raise `dipdup.exceptions.ProjectPackageError` here so we get a nicer help message.

droserasprout reviewed (changes requested):
Your last comment shows such a great amount of testing done! 🥰 I think, there's nothing critical. Yes, running commands in random order on multiple databases sometimes confuse aerich, but all real cases works perfectly.

I have just a few nitpicks. Then make sure that `make all docs` command passes, add yourself to the `thanks.md` file, and let's merge it!

bigherc18 committed:
Minor fixes for database migrations with aerich

bigherc18 committed:
Changes by running `make docs`

bigherc18 committed:
Merge branch 'next' into feat/add-db-migration

bigherc18 commented:
> I think, there's nothing critical. Yes, running commands in random order on multiple databases sometimes confuse aerich, but all real cases works perfectly.

Thank you for your feedback but I just want to make sure I explained my concerns clear enough for you

1. Multiple databases not supported by migrations

TLDR; if a user uses multiple databases at the same time, migrations is not supported

I see dipdup makes it easy for users to swap databases by just combining different config files with multiple `-c` args, the problem is: if a user swaps his database after doing some migrations in the first one, `aerich` is now really confused and I don't even know what will happen, the `aerich` table will be empty (or contains migrations history for the second database) but the `migrations` folder still contains files from the first database

I think this worth at least a note in the docs 

3. CLI handing

Here's a little video demonstrating the issue:

![demo](https://github.com/user-attachments/assets/fbf0f9e1-3daf-450f-a575-37af993aefe8)
